packages/oauth/src/state-stores/file-state-store.ts

```diff
@@ -0,0 +1,108 @@
+import { homedir } from 'os';
+import fs from 'fs';
+import path from 'path';
+import { randomUUID } from 'crypto';
+import { StateStore, StateObj } from './interface';
+import { InstallURLOptions } from '../install-url-options';
+import { InvalidStateError } from '../errors';
+
+export interface FileStateStoreArgs {
+  stateExpirationSeconds?: number;
+  baseDir?: string;
+}
+
+export class FileStateStore implements StateStore {
+  private baseDir: string;
+
+  private stateExpirationSeconds: number;
+
+  public constructor(args: FileStateStoreArgs) {
+    this.baseDir = args.baseDir !== undefined ?
+      args.baseDir :
+      `${homedir()}/.bolt-js-oauth-states`;
+    this.stateExpirationSeconds = args.stateExpirationSeconds !== undefined ?
+      args.stateExpirationSeconds :
+      600;
+  }
+
+  public async generateStateParam(
+    installOptions: InstallURLOptions,
+    now: Date,
+  ): Promise<string> {
+    const state = randomUUID();
+    const source: StateObj = {
+      installOptions,
+      now,
+    };
+    this.writeToFile(state, source);
+    return state;
+  }
+
+  public async verifyStateParam(
+    now: Date,
+    state: string,
+  ): Promise<InstallURLOptions> {
+    try {
+      if (this.findFile(state)) {
+        // decode the state using the secret
+        let decoded: StateObj | undefined;
+        try {
+          decoded = this.readFile(state);
+        } catch (e) {
+          const message = `Failed to load the data represented by the state parameter (error: ${e})`;
+          throw new InvalidStateError(message);
+        }
+        if (decoded !== undefined) {
+          // Check if the state value is not too old
+          const generatedAt = new Date(decoded.now);
+          const passedSeconds = Math.floor(
+            (now.getTime() - generatedAt.getTime()) / 1000,
+          );
+          if (passedSeconds > this.stateExpirationSeconds) {
+            throw new InvalidStateError('The state value is already expired');
+          }
+          // return installOptions
+          return decoded.installOptions;
+        }
+      }
+    } finally {
```

Comment:
If I understand this correctly, the file corresponding to a specific state value will be deleted in the case that it exists but reading it returned undefined (which, as it stands, will only happen in case of a file reading, JSON parsing, or string casting error), or, if the state is expired.

Is this by design? To prevent any potential state tampering / hacking / investigation? Is my understanding of the intention here correct that, as soon as there is any issue in decoding or reading a state value, the state is invalidated and removed and thus the user would need to restart the installation dance?

Reply:
You understanding here is correct. Regardless of the results of this method call, verification with the same state value should not be retried.